File: September/2026-09-03/813/813-largest-sum-of-avg.py
# 813-largest-sum-of-avg.py;leetcode;help;sum-prefix

class Solution:
    def largestSumOfAverages(self, nums: List[int], k: int) -> float:
        n = len(nums)
        prefix = [0] * (n + 1)

        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]

        dp = [0.0] * (n + 1)

        for i in range(1, n + 1):
            dp[i] = prefix[i] / i

        for groups in range(2, k + 1):
            new_dp = [0.0] * (n + 1)

            for i in range(groups, n + 1):
                for j in range(groups - 1, i):
                    average = (prefix[i] - prefix[j]) / (i - j)

                    new_dp[i] = max(
                        new_dp[i],
                        dp[j] + average
                    )

            dp = new_dp

        return dp[n]

File: September 2026/2026-09-03/813-largest-sum-of-avg.cpp
// 813-largest-sum-of-avg.cpp;prefixsum;help;leetcode

class Solution {
public:
    double largestSumOfAverages(vector<int>& nums, int k) {
        int n = nums.size();
        vector<double> prefix(n + 1, 0);

        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        vector<double> dp(n + 1);
        for (int i = 1; i <= n; i++) {
            dp[i] = prefix[i] / i;
        }

        for (int groups = 2; groups <= k; groups++) {

            vector<double> newDp(n + 1, 0);
            for (int i = groups; i <= n; i++) {
                for (int j = groups - 1; j < i; j++) {
                    double average = (prefix[i] - prefix[j]) / (i - j);

                    newDp[i] = max(newDp[i],dp[j] + average);
                }
            }

            dp = newDp;
        }

        return dp[n];
    }
};